xbmc/pvr/filesystem/PVRGUIDirectory.cpp

```diff
@@ -458,6 +458,109 @@ bool CPVRGUIDirectory::GetChannelGroupsDirectory(bool bRadio,
   return false;
 }
 
+namespace
+{
+std::shared_ptr<CPVRChannelGroupMember> GetLastWatchedChannelGroupMember(
+    const std::shared_ptr<CPVRChannel>& channel)
+{
+  const int lastGroupId{channel->LastWatchedGroupId()};
+  if (lastGroupId != PVR_GROUP_ID_UNNKOWN)
+  {
+    const std::shared_ptr<CPVRChannelGroup> lastGroup{
+        CServiceBroker::GetPVRManager().ChannelGroups()->GetByIdFromAll(lastGroupId)};
+    if (lastGroup && !lastGroup->IsHidden() && !lastGroup->IsDeleted())
+      return lastGroup->GetByUniqueID(channel->StorageId());
+  }
+  return {};
+}
+
+std::shared_ptr<CPVRChannelGroupMember> GetFirstMatchingGroupMember(
+    const std::shared_ptr<CPVRChannel>& channel)
+{
+  CPVRChannelGroups* groups{
+      CServiceBroker::GetPVRManager().ChannelGroups()->Get(channel->IsRadio())};
+  if (groups)
+  {
+    const std::vector<std::shared_ptr<CPVRChannelGroup>> channelGroups{
+        groups->GetMembers(true /* exclude hidden */)};
+
+    for (const auto& channelGroup : channelGroups)
+    {
+      if (channelGroup->IsDeleted())
+        continue;
+
+      const std::shared_ptr<CPVRChannelGroupMember> groupMember{
+          channelGroup->GetByUniqueID(channel->StorageId())};
+      if (groupMember)
+        return groupMember;
+    }
+  }
+  return {};
+}
+
+std::vector<std::shared_ptr<CPVRChannelGroupMember>> GetChannelGroupMembers(
+    const CPVRChannelsPath& path)
+{
+  const std::string& groupName{path.GetGroupName()};
+
+  std::shared_ptr<CPVRChannelGroup> group;
+  if (path.IsHiddenChannelGroup()) // hidden channels from the 'all channels' group
+  {
+    group = CServiceBroker::GetPVRManager().ChannelGroups()->GetGroupAll(path.IsRadio());
+  }
+  else if (groupName == "*") // all channels across all groups
+  {
+    group = CServiceBroker::GetPVRManager().ChannelGroups()->GetGroupAll(path.IsRadio());
+    if (group)
+    {
+      std::vector<std::shared_ptr<CPVRChannelGroupMember>> result;
+
+      const std::vector<std::shared_ptr<CPVRChannelGroupMember>> allGroupMembers{
+          group->GetMembers(CPVRChannelGroup::Include::ONLY_VISIBLE)};
+      for (const auto& allGroupMember : allGroupMembers)
+      {
+        std::shared_ptr<CPVRChannelGroupMember> member{
+            GetLastWatchedChannelGroupMember(allGroupMember->Channel())};
+        if (member)
+        {
+          result.emplace_back(member);
+          continue; // Process next 'All channels' group member.
+        }
+
+        if (group->IsHidden())
+        {
+          // Very special case. 'All channels' group is hidden. Let's see what we get iterating all
+          // non-hidden / non-deleted groups. We must not return any all channels group members,
```

Comment:
Minor cahnge `all channels` to `'All channels'`

Reply:
done